sdk/helper/template/template.go

```diff
@@ -0,0 +1,140 @@
+	return func(up *StringTemplate) error {
+		up.rawTemplate = rawTemplate
+		return nil
+	}
+}
+
+// Function allows the user to specify functions for use in the template. If the name provided is a function that
+// already exists in the function map, this will override the previously specified function.
+func Function(name string, f interface{}) Opt {
+	return func(up *StringTemplate) error {
+		if name == "" {
+			return fmt.Errorf("missing function name")
+		}
+		if f == nil {
+			return fmt.Errorf("missing function")
+		}
+		up.funcMap[name] = f
+		return nil
+	}
+}
+
+// StringTemplate creates strings based on the provided template.
+// This uses the go templating language, so anything that adheres to that language will function in this struct.
+// There are several custom functions available for use in the template:
+// - random (rand)
+//   - Randomly generated characters. This uses the charset specified in RandomCharset. Must include a length.
+//     Example: {{ rand 20 }}
+// - truncate (trunc)
+//   - Truncates the previous value to the specified length. Must include a maximum length.
+//     Example: {{ .DisplayName | truncate 10 }}
+// - truncate_sha256 (trunc_sha256)
+//   - Truncates the previous value to the specified length. If the original length is greater than the length
+//     specified, the remaining characters will be sha256 hashed and appended to the end. The hash will be only the first 8 characters The maximum length will
+//     be no longer than the length specified.
+//     Example: {{ .DisplayName | truncate_sha256 30 }}
+// - uppercase (upper)
+//   - Uppercases the previous value.
+//     Example: {{ .RoleName | uppercase }}
+// - lowercase (lower)
+//   - Lowercases the previous value.
+//     Example: {{ .DisplayName | lowercase }}
+// - replace
+//   - Performs a string find & replace
+//     Example: {{ .DisplayName | replace - _ }}
+// - sha256
+//   - SHA256 hashes the previous value.
+//     Example: {{ .DisplayName | sha256 }}
+// - timestamp (now_seconds)
+//   - Provides the current unix time in seconds.
+//     Example: {{ now_seconds}}
+// - now_nano
+//   - Provides the current unix time in nanoseconds.
+//     Example: {{ now_nano}}
+// - uuid
+//   - Generates a UUID
+//     Example: {{ uuid }}
+type StringTemplate struct {
+	rawTemplate string
+	tmpl        *template.Template
+	funcMap     template.FuncMap
+}
+
+// NewTemplate creates a StringTemplate. No arguments are required
+// as this has reasonable defaults for all values.
+// The default template is specified in the DefaultTemplate constant.
+func NewTemplate(opts ...Opt) (up StringTemplate, err error) {
+	up = StringTemplate{
+		funcMap: map[string]interface{}{
+			"random":          base62.Random,
+			"rand":            base62.Random,
+			"truncate":        truncate,
+			"trunc":           truncate,
+			"truncate_sha256": truncateSHA256,
+			"trunc_sha256":    truncateSHA256,
+			"uppercase":       uppercase,
```

Comment:
I wonder whether it's necessary to expose two versions of the same func. There's extensive documentation on the godoc, and there will likely be just as much documentation on our site on their usage as well.

Reply:
If I were to choose one I'd rather keep the more explicit one. I originally did this so "power users" could use shortened templates but I'm open to not having duplicates as well.

Reply:
I do see your point of making things easier for power users, but I also like the idea of sticking with one way to do the same thing even if it ends up being a bit longer.